src/haddock/libs/libplots.py

```diff
@@ -523,29 +561,317 @@ def scatter_plot_handler(capri_filename, cl_rank, png, dpi):
+        cols=number_of_cols,
+        shared_xaxes=shared_xaxes,
+        shared_yaxes=shared_yaxes,
+        vertical_spacing=(0.4 / number_of_rows),
+        horizontal_spacing=(0.3 / number_of_cols),
+        )
+    for i, sub_fig in enumerate(plots):
+        col_index = int((i % number_of_cols) + 1)
+        row_index = int(np.floor(i / number_of_cols) + 1)
+        # hide legend of plots except one
+        if i != 0:
+            sub_fig.for_each_trace(lambda trace: trace.update(showlegend=False))
+        fig.add_traces(sub_fig.data, rows=row_index, cols=col_index)
+        fig.update_yaxes(
+            title_text=sub_fig.layout.yaxis.title.text,
+            row=row_index,
+            col=col_index,
+            title_standoff=5,
+            automargin=True,
+            )
+        # x title only on the last row
+        if shared_xaxes == "all":
+            row_index = number_of_rows
+        fig.update_xaxes(
+            title_text=sub_fig.layout.xaxis.title.text,
+            row=row_index,
+            col=col_index,
+            title_standoff=5,
+            automargin=True,
+            )
+        legend_title_text = sub_fig.layout.legend.title.text
+    fig.update_layout(
+        title_text=plot_title,
+        legend_title_text=legend_title_text,
+        height=height * number_of_rows,
+        width=width * number_of_cols,
+        )
+    return fig
+
+
+def find_best_struct(ss_file, number_of_struct=4):
+    """
+    Find best structures.
+
+    It inspects model-cluster-ranking recorded in capri_ss.tsv file and finds
+    the best models (models with lower ranks). By default, it selects the 4 best
+    models.
+
+    Parameters
+    ----------
+    ss_file : path
+        path to capri_ss.tsv
+    number_of_struct: int
+        number of models with lower model-cluster-ranking
+
+    Returns
+    -------
+    best_struct_df : pandas DataFrame
+        DataFrame of best structures
+    """
+    dfss = read_capri_table(ss_file)
+    dfss = dfss.sort_values(by=["cluster-id", "model-cluster-ranking"])
+    # TODO need a check for "Unclustered"
+    # possible number of structures, can be different per each cluster
+    max_number_of_struct = dfss.groupby("cluster-id").count()["model-cluster-ranking"].min()
+    number_of_struct = min(number_of_struct, max_number_of_struct)
+    best_struct_df = dfss.groupby("cluster-id").head(number_of_struct).copy()
+    number_of_cluster = len(best_struct_df["cluster-id"].unique())
+    col_names = [
+        f"Nr {number + 1} best structure" for number in range(number_of_struct)
+        ] * number_of_cluster
+    best_struct_df = best_struct_df.assign(Structure=col_names)
+    best_struct_df = best_struct_df.pivot_table(
+        index=["cluster-id"],
+        columns=["Structure"],
+        values="model",
+        aggfunc=lambda x: x,
+        )
+    best_struct_df.reset_index(inplace=True)
+    best_struct_df.rename(columns={"cluster-id": "Cluster ID"}, inplace=True)
```

Comment:
could you add some comments here and there to help clarifying the procedure?

Reply:
sure, I added comments. Let me know if something is still unclear.